### PR TITLE
[feature](Nereids): use Multi join to rearrange join to eliminate cross join by using predicate.

### PR DESCRIPTION
**Pull request by jackwener**
# Proposed changes

Issue Number: close #xxx

## Problem summary

use Multi join to rearrange join to eliminate cross join

- convert `join cluster` to `multijoin`
- convert `multijoin` to `join` (eliminate cross join)

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MultiJoin.java [hunk 1]
```diff
@@ -17,189 +17,170 @@
...
+    // Push predicates into it.
+    // But joinFilter shouldn't contain predicate which just contains one predicate like `T.key > 1`.
+    // Because these predicate should be pushdown.
+    private final List<Expression> joinFilter;
+    // MultiJoin just contains one OUTER/SEMI/ANTI.
+    private final Optional<JoinType> onlyJoinType;
+    // When contains one OUTER/SEMI/ANTI join, keep separately its condition.
+    private final List<Expression> notInnerJoinConditions;
+
+    // private final List<@Nullable List<NamedExpression>> projFields;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MultiJoin.java:73]: why comment this？

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MultiJoin.java [hunk 2]
```diff
@@ -17,189 +17,170 @@
...
+
+    public MultiJoin(List<Plan> inputs, List<Expression> joinFilter, Optional<JoinType> onlyJoinType,
+            List<Expression> notInnerJoinConditions) {
+        super(PlanType.LOGICAL_MULTI_JOIN, inputs.toArray(new Plan[0]));
+        this.joinFilter = joinFilter;
+        this.onlyJoinType = onlyJoinType;
+        this.notInnerJoinConditions = notInnerJoinConditions;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MultiJoin.java:80]: Objects.isNotNull?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MultiJoin.java [hunk 3]
```diff
@@ -17,189 +17,170 @@
...
-            return Optional.of(PlanUtils.filterOrSelf(conjunctsKeepInFilter, root));
-        }
-        return Optional.empty();
+
+    // Push predicates into it.
+    // But joinFilter shouldn't contain predicate which just contains one predicate like `T.key > 1`.
+    // Because these predicate should be pushdown.
+    private final List<Expression> joinFilter;
+    // MultiJoin just contains one OUTER/SEMI/ANTI.
+    private final Optional<JoinType> onlyJoinType;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MultiJoin.java:69]: i think joinType is enough

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 4]
```diff
@@ -37,21 +56,296 @@
...
         }).toRule(RuleType.REORDER_JOIN);
     }
+
+    /**
+     * Recursively convert to
+     * {@link LogicalJoin} or
+     * {@link LogicalFilter}--{@link LogicalJoin}
+     * --> {@link MultiJoin}
+     */
+    public Plan joinToMultiJoin(Plan plan) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:84]: should we put this function into JoinUtils?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 5]
```diff
@@ -37,21 +56,296 @@
...
+
+        List<Plan> inputs = Lists.newArrayList();
+        List<Expression> joinFilter = Lists.newArrayList();
+        List<Expression> notInnerJoinConditions = Lists.newArrayList();
+
+        LogicalJoin<?, ?> join;
+        if (plan instanceof LogicalFilter) {
+            LogicalFilter<?> filter = (LogicalFilter<?>) plan;
+            joinFilter.addAll(ExpressionUtils.extractConjunction(filter.getPredicates()));
+            join = (LogicalJoin<?, ?>) filter.child();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:99]: we have a Implicit rely on MergeFilter. Add this info on class comment~

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 6]
```diff
@@ -37,21 +56,296 @@
...
+    }
+
+    /**
+     * Returns whether an input can be merged without changing semantics.
+     *
+     * @param input input into a MultiJoin or (GroupPlan|LogicalFilter)
+     * @param changeLeft generate nullable or left not exist.
+     * @return true if the input can be combined into a parent MultiJoin
+     */
+    private static boolean canCombine(Plan input, boolean changeLeft) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:290]:
```suggestion
    private static boolean canCombine(Plan input, boolean outputNullable) {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 7]
```diff
@@ -37,21 +56,296 @@
...
+        newInputs.add(join);
+        newInputs.addAll(candidates.stream().filter(plan -> !join.right().equals(plan)).collect(Collectors.toList()));
+
+        joinFilter.removeAll(join.getHashJoinConjuncts());
+        joinFilter.removeAll(join.getOtherJoinConjuncts());
+        return multiJoinToJoin(new MultiJoin(
+                newInputs,
+                joinFilter,
+                Optional.empty(),
+                ExpressionUtils.EMPTY_CONDITION));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:280]: could eliminate this recursion？

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 8]
```diff
@@ -37,21 +56,296 @@
...
+                Preconditions.checkState(!exprInputSlots.isEmpty());
+
+                if (left.getOutputSet().containsAll(exprInputSlots)) {
+                    leftFilter.add(expr);
+                } else if (right.getOutputSet().containsAll(exprInputSlots)) {
+                    rightFilter.add(expr);
+                } else if (multiJoin.getOutputSet().containsAll(exprInputSlots)) {
+                    remainingFilter.add(expr);
+                } else {
+                    throw new RuntimeException("invalid expression");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:252]: add more info to exception, and do log

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MultiJoin.java (hunk 1 above)
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MultiJoin.java:73]: current it's useless

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java (hunk 6 above)
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:290]: It can be SEMI/ANTI

**Review by morrySnow — APPROVED** [review]

**Pull request merged**